# Inhibitory vs Exitatory

### creline reporter positive only

In [60]:
%matplotlib inline

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Data load & check

In [62]:
cell_data = pd.read_csv('cell_data.csv')
cell_data.head()

line_name  specimen__id                         specimen__name  \
0   Scnn1a-Tg2-Cre     313862022    Scnn1a-Tg2-Cre;Ai14-167081.03.02.01   
1   Rorb-IRES2-Cre     479508018  Rorb-IRES2-Cre-D;Ai14-196662.03.01.01   
2  Htr3a-Cre_NO152     442627661   Htr3a-Cre_NO152;Ai14-178904.06.01.01   
3     Cux2-CreERT2     487358945      Cux2-CreERT2;Ai14-214155.05.01.01   
4   Scnn1a-Tg3-Cre     469803127    Scnn1a-Tg3-Cre;Ai14-180215.05.02.01   

  specimen__hemisphere  structure__id               structure__name  \
0                right            721  Primary visual area, layer 4   
1                 left            721  Primary visual area, layer 4   
2                 left      312782566       Anterior area, layer 6a   
3                 left            778  Primary visual area, layer 5   
4                right            778  Primary visual area, layer 5   

  structure__acronym  structure_parent__id structure_parent__acronym  \
0              VISp4                   385                      VISp   
1              VISp4                   385                      VISp   
2             VISa6a             312782546                      VISa   
3              VISp5                   385                      VISp   
4              VISp5                   385                      VISp   

  structure__layer ...    cell_reporter_status  m__glif  \
0                4 ...   cre reporter positive        0   
1                4 ...   cre reporter positive        2   
2               6a ...   cre reporter positive        0   
3                5 ...   cre reporter positive        5   
4                5 ...   cre reporter positive        5   

   m__biophys_perisomatic  m__biophys_all_active  tag__apical  \
0                       1                      0       intact   
1                       0                      0       intact   
2                       0                      0          NaN   
3                       0                      0       intact   
4                       0                      0       intact   

   tag__dendrite_type                            morph_thumb_path  \
0               spiny  /api/v2/well_known_file_download/487659609   
1               spiny                                         NaN   
2              aspiny                                         NaN   
3               spiny                                         NaN   
4               spiny                                         NaN   

                             ephys_thumb_path  \
0  /api/v2/well_known_file_download/487623088   
1  /api/v2/well_known_file_download/491384619   
2  /api/v2/well_known_file_download/487610447   
3  /api/v2/well_known_file_download/487361512   
4  /api/v2/well_known_file_download/487635954   

                 ephys_inst_thresh_thumb_path   id  
0  /api/v2/well_known_file_download/491376069  606  
1  /api/v2/well_known_file_download/491384621  522  
2  /api/v2/well_known_file_download/491377923   81  
3  /api/v2/well_known_file_download/491767690   38  
4  /api/v2/well_known_file_download/491376335  671  

[5 rows x 44 columns]

In [64]:
len(cell_data)

692

## e_phys data load

In [65]:
ephy_data = pd.read_csv('ephys_features.csv')
ephy_data.head()

Unnamed: 0  adaptation    avg_isi  electrode_0_pa  f_i_curve_slope  \
0           0         NaN  44.797500       11.978124         0.084259   
1           1    0.032340  55.895000      -55.964379         0.280000   
2           2    0.162880  59.473000        5.602500         0.091032   
3           3    0.113440  86.349000       -9.024376         0.204990   
4           4    0.030623  72.309615       -6.614375         0.251415   

   fast_trough_t_long_square  fast_trough_t_ramp  fast_trough_t_short_square  \
0                   0.101075            5.280735                    0.008300   
1                   0.094755            1.833668                    0.008380   
2                   0.048620            5.704792                    0.008326   
3                   0.143820            4.080805                    0.004870   
4                   0.173805            3.753390                    0.008297   

   fast_trough_v_long_square  fast_trough_v_ramp    ...      trough_t_ramp  \
0                 -59.531254          -57.697918    ...           5.280735   
1                 -50.250000          -52.593754    ...           1.865633   
2                 -51.531250          -53.104169    ...           5.715375   
3                 -43.968754          -46.072919    ...           4.112943   
4                 -51.906254          -49.500000    ...           3.785528   

  trough_t_short_square trough_v_long_square trough_v_ramp  \
0              0.510604           -59.531254    -57.697918   
1              0.500193           -54.031254    -57.385419   
2              0.402444           -51.625004    -55.322919   
3              0.620560           -51.375000    -51.489586   
4              0.358858           -58.937500    -54.552086   

   trough_v_short_square  upstroke_downstroke_ratio_long_square  \
0             -78.536465                               2.009734   
1             -77.750005                               3.042933   
2             -75.195314                               3.758079   
3             -77.929691                               3.617618   
4             -74.912503                               4.290042   

   upstroke_downstroke_ratio_ramp  upstroke_downstroke_ratio_short_square  \
0                        1.980484                                1.957409   
1                        3.517684                                3.274181   
2                        3.308425                                3.694207   
3                        3.732209                                3.773257   
4                        4.340203                                4.628210   

   vm_for_sag      vrest  
0  -96.781250 -77.704921  
1 -101.000000 -76.928478  
2  -89.968758 -74.231581  
3  -94.718758 -77.662368  
4  -90.031250 -74.220647  

[5 rows x 57 columns]

In [70]:
len(ephy_data)

813

### merge specimen id

In [75]:
mdata = pd.merge(cell_data, ephy_data, how='inner', left_on=['specimen__id'], right_on=['specimen_id'])

In [84]:
len(mdata)

692

In [85]:
mdata.head()

line_name  specimen__id                         specimen__name  \
0   Scnn1a-Tg2-Cre     313862022    Scnn1a-Tg2-Cre;Ai14-167081.03.02.01   
1   Rorb-IRES2-Cre     479508018  Rorb-IRES2-Cre-D;Ai14-196662.03.01.01   
2  Htr3a-Cre_NO152     442627661   Htr3a-Cre_NO152;Ai14-178904.06.01.01   
3     Cux2-CreERT2     487358945      Cux2-CreERT2;Ai14-214155.05.01.01   
4   Scnn1a-Tg3-Cre     469803127    Scnn1a-Tg3-Cre;Ai14-180215.05.02.01   

  specimen__hemisphere  structure__id               structure__name  \
0                right            721  Primary visual area, layer 4   
1                 left            721  Primary visual area, layer 4   
2                 left      312782566       Anterior area, layer 6a   
3                 left            778  Primary visual area, layer 5   
4                right            778  Primary visual area, layer 5   

  structure__acronym  structure_parent__id structure_parent__acronym  \
0              VISp4                   385                      VISp   
1              VISp4                   385                      VISp   
2             VISa6a             312782546                      VISa   
3              VISp5                   385                      VISp   
4              VISp5                   385                      VISp   

  structure__layer    ...      trough_t_ramp  trough_t_short_square  \
0                4    ...           2.570595               0.065430   
1                4    ...           2.515478               0.300489   
2               6a    ...          12.013955               0.596460   
3                5    ...           2.996695               0.747021   
4                5    ...           2.187125               0.775278   

   trough_v_long_square  trough_v_ramp  trough_v_short_square  \
0            -50.875000     -51.156254             -64.531250   
1            -53.625004     -57.239587             -72.273445   
2            -48.468750     -57.958335             -78.447918   
3            -54.250000     -57.041669             -79.081253   
4            -50.812500     -50.296879             -78.800003   

   upstroke_downstroke_ratio_long_square  upstroke_downstroke_ratio_ramp  \
0                               5.261166                        5.386440   
1                               4.957590                        4.685049   
2                               4.843871                        3.656743   
3                               4.815331                        4.534321   
4                               4.756384                        3.489439   

   upstroke_downstroke_ratio_short_square  vm_for_sag      vrest  
0                                5.949575  -70.906250 -63.574978  
1                                4.538613  -92.781250 -72.237417  
2                                2.261516  -83.687500 -78.260158  
3                                4.969444  -92.218758 -77.838111  
4                                4.384974  -97.968758 -76.375813  

[5 rows x 101 columns]

In [79]:
mdata.describe()

/home/hyunsu/anaconda3/lib/python3.5/site-packages/numpy/lib/function_base.py:3823: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


specimen__id  structure__id  structure_parent__id  \
count  6.920000e+02   6.920000e+02          6.920000e+02   
mean   4.558896e+08   1.310858e+07          4.520371e+06   
std    5.887130e+07   6.271990e+07          3.735443e+07   
min    3.138614e+08   3.300000e+01          3.850000e+02   
25%    4.716713e+08   7.210000e+02          3.850000e+02   
50%    4.806734e+08   7.780000e+02          3.850000e+02   
75%    4.873108e+08   7.780000e+02          3.850000e+02   
max    5.096209e+08   3.127826e+08          3.127826e+08   

       nr__max_euclidean_distance  nr__number_stems  nr__number_bifurcations  \
count                2.590000e+02        259.000000               259.000000   
mean                 1.035079e+04          6.648649                25.698842   
std                  1.595810e+05          2.157578                11.732898   
min                  3.458333e+01          2.000000                 3.000000   
25%                           NaN               NaN                      NaN   
50%                           NaN               NaN                      NaN   
75%                           NaN               NaN                      NaN   
max                  2.568642e+06         14.000000                78.000000   

       nr__average_contraction  nr__average_parent_daughter_ratio  \
count               259.000000                         259.000000   
mean                  0.896209                           0.768706   
std                   0.027687                           0.068536   
min                   0.697496                           0.446585   
25%                        NaN                                NaN   
50%                        NaN                                NaN   
75%                        NaN                                NaN   
max                   0.958597                           1.021573   

          nrwkf__id     erwkf__id     ...      trough_t_ramp  \
count  2.430000e+02  6.920000e+02     ...         655.000000   
mean   4.950256e+08  4.954025e+08     ...           6.212352   
std    9.113933e+06  6.378012e+06     ...           4.736947   
min    3.979054e+08  4.870963e+08     ...           0.285193   
25%             NaN  4.912012e+08     ...                NaN   
50%             NaN  4.961962e+08     ...                NaN   
75%             NaN  4.966043e+08     ...                NaN   
max    5.143516e+08  5.156427e+08     ...          32.018185   

       trough_t_short_square  trough_v_long_square  trough_v_ramp  \
count             692.000000            692.000000     655.000000   
mean                0.492196            -56.667361     -55.603191   
std                 0.350256              5.211311       3.893259   
min                 0.003245            -75.656250     -73.312508   
25%                 0.274357            -59.820313            NaN   
50%                 0.449953            -56.156250            NaN   
75%                 0.615760            -53.304690            NaN   
max                 1.875384            -41.593750     -37.791668   

       trough_v_short_square  upstroke_downstroke_ratio_long_square  \
count             692.000000                             692.000000   
mean              -73.877066                               2.471768   
std                 4.900303                               1.044234   
min               -86.492189                               0.862726   
25%               -77.428129                               1.427780   
50%               -74.209377                               2.462087   
75%               -70.778128                               3.352583   
max               -59.125004                               5.261166   

       upstroke_downstroke_ratio_ramp  upstroke_downstroke_ratio_short_square  \
count                      655.000000                              692.000000   
mean                         2.462993                                2.534390   
std                          1.100193    

In [81]:
mdata.dtypes

line_name                                     object
specimen__id                                   int64
specimen__name                                object
specimen__hemisphere                          object
structure__id                                  int64
structure__name                               object
structure__acronym                            object
structure_parent__id                           int64
structure_parent__acronym                     object
structure__layer                              object
nr__max_euclidean_distance                   float64
nr__number_stems                             float64
nr__number_bifurcations                      float64
nr__average_contraction                      float64
nr__average_parent_daughter_ratio            float64
nrwkf__id                                    float64
erwkf__id                                      int64
ef__upstroke_downstroke_ratio_long_square    float64
ef__adaptation                               f

### Whole data fearture 101 after merge, observation 692

In [82]:
data_cre = mdata[data['cell_reporter_status'] == 'cre reporter positive']
len(data_cre)

615

### cre positive cell 615

### cre reporter posivitive cells tagging.

In [89]:
def dis_EI(cre_line): # labelling for excitory vs inbihitory cre line
    output = []
    for item in cre_line:
        if (item == 'Pvalb-IRES-Cre' or item == 'Sst-IRES-Cre' or
            item == 'Htr3a-Cre_NO152' or item == 'Vip-IRES-Cre' or
            item == 'Gad2-IRES-Cre'):
                output = output + ['I']
        else:
            output = output + ['E']
    return output
ExIn = []
ExIn = dis_EI(data_cre['line_name']) # this is python list!!
ExIn_pd = pd.DataFrame({'ExIn': ExIn}) # making pandas frame
data_cre = data_cre.join(ExIn_pd)


ValueError: columns overlap but no suffix specified: Index(['ExIn'], dtype='object')

## subsetting data - only electrophysiological component for comparision

In [110]:
data_sub = data_cre.iloc[:, 0:1]
# data_sub = data_sub.join(data_cre.iloc[:, 17:27]) 중복인듯
data_sub = data_sub.join(data_cre.iloc[:, 45:])
data_sub.describe()

/home/hyunsu/anaconda3/lib/python3.5/site-packages/numpy/lib/function_base.py:3823: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


adaptation     avg_isi  electrode_0_pa  f_i_curve_slope  \
count  502.000000  573.000000      613.000000       615.000000   
mean     0.046840   60.449486       -5.876237         0.391070   
std      0.072595   50.028697       35.854419         0.392468   
min     -0.204913    3.828333     -166.588104        -0.015048   
25%           NaN         NaN             NaN         0.145817   
50%           NaN         NaN             NaN         0.220765   
75%           NaN         NaN             NaN         0.498809   
max      0.586461  505.870000      256.913130         2.382667   

       fast_trough_t_long_square  fast_trough_t_ramp  \
count                 615.000000          579.000000   
mean                    0.206037            6.113492   
std                     0.308923            4.628041   
min                     0.003920            0.285193   
25%                     0.047010                 NaN   
50%                     0.095490                 NaN   
75%                     0.177127                 NaN   
max                     1.774100           28.141403   

       fast_trough_t_short_square  fast_trough_v_long_square  \
count                  615.000000                 615.000000   
mean                     0.075610                 -54.326882   
std                      0.253973                   6.645251   
min                      0.003208                 -75.656250   
25%                      0.008172                 -59.250004   
50%                      0.008321                 -53.625004   
75%                      0.008429                 -49.171877   
max                      1.027960                 -39.437504   

       fast_trough_v_ramp  fast_trough_v_short_square     ...      \
count          579.000000                  615.000000     ...       
mean           -52.893645                  -59.724947     ...       
std              5.104135                    6.344713     ...       
min            -67.583333                  -78.525008     ...       
25%                   NaN                  -65.075002     ...       
50%                   NaN                  -59.026044     ...       
75%                   NaN                  -54.235420     ...       
max            -37.791668                  -46.382814     ...       

       trough_t_ramp  trough_t_short_square  trough_v_long_square  \
count     579.000000             615.000000            615.000000   
mean        6.128833               0.502785            -56.598224   
std         4.620625               0.353485              5.150431   
min         0.285193               0.003245            -75.656250   
25%              NaN               0.292018            -59.828127   
50%              NaN               0.455093            -56.062500   
75%              NaN               0.630304            -53.296877   
max        28.141403               1.875384            -41.593750   

       trough_v_ramp  trough_v_short_square  \
count     579.000000             615.000000   
mean      -55.544948             -73.907549   
std         3.841003               4.921210   
min       -67.583333             -86.492189   
25%              NaN             -77.460942   
50%              NaN             -74.291672   
75%              NaN             -70.803129   
max       -37.791668             -59.125004   

       upstroke_downstroke_ratio_long_square  upstroke_downstroke_ratio_ramp  \
count                             615.000000                      579.000000   
mean                                2.470953                        2.462301   
std                                 1.037507                        1.090527   
min                                 0.862726                        0.724582   
25%                                 1.447692                             NaN   
50%                                 2.459218                             NaN   
75%                                 3.346059                             NaN   
max                   

In [111]:
data_sub.dtypes

line_name                                  object
adaptation                                float64
avg_isi                                   float64
electrode_0_pa                            float64
f_i_curve_slope                           float64
fast_trough_t_long_square                 float64
fast_trough_t_ramp                        float64
fast_trough_t_short_square                float64
fast_trough_v_long_square                 float64
fast_trough_v_ramp                        float64
fast_trough_v_short_square                float64
has_burst                                  object
has_delay                                  object
has_pause                                  object
id_y                                        int64
input_resistance_mohm                     float64
latency                                   float64
peak_t_long_square                        float64
peak_t_ramp                               float64
peak_t_short_square                       float64


### value counts

In [112]:
line_name_freq = pd.DataFrame(data_sub['line_name'].value_counts())
line_name_freq

line_name
Pvalb-IRES-Cre          139
Sst-IRES-Cre            108
Htr3a-Cre_NO152          78
Rorb-IRES2-Cre           70
Rbp4-Cre_KL100           49
Cux2-CreERT2             44
Nr5a1-Cre                40
Scnn1a-Tg3-Cre           33
Ntsr1-Cre_GN220          19
Scnn1a-Tg2-Cre           15
Gad2-IRES-Cre            11
Slc17a6-IRES-Cre          3
Ntsr1-Cre                 2
Ctgf-2A-dgCre             2
Vip-IRES-Cre              2

In [113]:
ExIn_freq = pd.DataFrame(data_sub['ExIn'].value_counts())
ExIn_freq

ExIn
I   308
E   242

In [114]:
data_sub_E = data_sub[data_sub['ExIn'] == 'E']
len(data_sub_E)

242

In [115]:
data_sub_I = data_sub[data_sub['ExIn'] == 'I']
len(data_sub_I)

308

In [116]:
data_sub_E.mean(0)

adaptation                                6.177148e-02
avg_isi                                   8.839027e+01
electrode_0_pa                           -3.536333e+00
f_i_curve_slope                           1.750761e-01
fast_trough_t_long_square                 2.482995e-01
fast_trough_t_ramp                        4.322476e+00
fast_trough_t_short_square                1.006834e-01
fast_trough_v_long_square                -4.956960e+01
fast_trough_v_ramp                       -4.998778e+01
fast_trough_v_short_square               -5.513997e+01
id_y                                      4.754228e+08
input_resistance_mohm                     1.797441e+02
latency                                   4.243054e+01
peak_t_long_square                        2.440881e-01
peak_t_ramp                               4.318287e+00
peak_t_short_square                       9.614212e-02
peak_v_long_square                        3.525994e+01
peak_v_ramp                               3.785871e+01
peak_v_sho

In [117]:
data_sub_I.mean(0)

adaptation                                4.260302e-02
avg_isi                                   4.524212e+01
electrode_0_pa                           -8.762788e+00
f_i_curve_slope                           4.365780e-01
fast_trough_t_long_square                 1.651378e-01
fast_trough_t_ramp                        6.338518e+00
fast_trough_t_short_square                6.690814e-02
fast_trough_v_long_square                -5.643415e+01
fast_trough_v_ramp                       -5.437987e+01
fast_trough_v_short_square               -6.162841e+01
id_y                                      4.699903e+08
input_resistance_mohm                     2.074441e+02
latency                                   3.078275e+01
peak_t_long_square                        1.632572e-01
peak_t_ramp                               6.336382e+00
peak_t_short_square                       6.298695e-02
peak_v_long_square                        2.313231e+01
peak_v_ramp                               2.469561e+01
peak_v_sho

In [118]:
data_sub_E.std(0)

adaptation                                6.882040e-02
avg_isi                                   4.802337e+01
electrode_0_pa                            3.729562e+01
f_i_curve_slope                           8.423935e-02
fast_trough_t_long_square                 3.239843e-01
fast_trough_t_ramp                        2.363944e+00
fast_trough_t_short_square                2.930403e-01
fast_trough_v_long_square                 3.641798e+00
fast_trough_v_ramp                        3.166979e+00
fast_trough_v_short_square                3.816167e+00
id_y                                      2.699841e+07
input_resistance_mohm                     7.215136e+01
latency                                   2.488024e+01
peak_t_long_square                        3.244019e-01
peak_t_ramp                               2.363933e+00
peak_t_short_square                       2.938449e-01
peak_v_long_square                        8.582452e+00
peak_v_ramp                               8.260338e+00
peak_v_sho

In [119]:
data_sub_I.std(0)

adaptation                                7.605492e-02
avg_isi                                   4.020753e+01
electrode_0_pa                            3.556419e+01
f_i_curve_slope                           3.706639e-01
fast_trough_t_long_square                 2.868775e-01
fast_trough_t_ramp                        4.747429e+00
fast_trough_t_short_square                2.390372e-01
fast_trough_v_long_square                 6.149367e+00
fast_trough_v_ramp                        5.173212e+00
fast_trough_v_short_square                5.701742e+00
id_y                                      2.954199e+07
input_resistance_mohm                     8.625156e+01
latency                                   6.946507e+01
peak_t_long_square                        2.867483e-01
peak_t_ramp                               4.747852e+00
peak_t_short_square                       2.396679e-01
peak_v_long_square                        1.105700e+01
peak_v_ramp                               1.260039e+01
peak_v_sho